### PR TITLE
[NT-2057] Updates showing project comments instead of updates

### PR DESCRIPTION
**Pull request by moderateepheezy**
# 📲 What

Fix issue where comments posted for a project update are reflecting on project comment section.

# 🤔 Why

We want to separate comments for project and comments for a project update, there for every comment posted on a project update should not reflect on project comments, and vice-versa.

# 👀 See

| Before 🐛 | After 🦋 |
| --- | --- |
|||

# ✅ Acceptance criteria

- [x] Confirm that posting a comment on a project update doesn't reflect on the project comment section.
- [x] Confirm that posting a comment on a project doesn't reflect on any of the project update comment section.
- [x] Confirm that when you retry a failed comment post on a project update, it doesn't reflect on project comment, and vice-versa.
- [ ] Verify that there's no regression on project comments section.
- [x] Confirm that pagination works as expected for both project comments and project update comments.

### CODE REVIEW

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift [hunk 1]
```diff
@@ -192,7 +192,7 @@ extension CommentsViewController {
   override func tableView(_: UITableView, didSelectRowAt indexPath: IndexPath) {
     guard let comment = self.dataSource.comment(at: indexPath) else { return }
 
-    self.viewModel.inputs.didSelectComment(comment)
+    self.viewModel.inputs.viewReplies(for: comment)
```
Comment (singhhari) [inline Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:195]: I think didSelectComment is the right name for the input. We usually follow the convention of the input being similarly named to the delegate method it is called in.

File: KsApi/GraphSchema.swift [hunk 2]
```diff
@@ -45,6 +45,10 @@ public func decodeBase64(_ input: String) -> String? {
     .flatMap { String(data: $0, encoding: .utf8) }
 }
 
+public func encodeToBase64(_ input: String) -> String {
```
Comment (singhhari) [inline KsApi/GraphSchema.swift:48]: alphabetize

File: Library/ViewModels/CommentsViewModel.swift [hunk 3]
```diff
@@ -216,13 +216,30 @@ public final class CommentsViewModel: CommentsViewModelType,
...
-    )
-    .takePairWhen(commentComposerDidSubmitText)
-    .map(unpack)
-    .flatMap(.concurrent(limit: concurrentCommentLimit), postCommentProducer)
+    let commentableId = projectOrUpdate
+      .flatMap { projectOrUpdate in
+        projectOrUpdate.ifLeft { project in
+          SignalProducer.init(value: project.graphID)
+        } ifRight: { update in
+          SignalProducer.init(value: encodeToBase64("FreeformPost-\(update.id)"))
```
Comment (singhhari) [inline Library/ViewModels/CommentsViewModel.swift:225]: What is the relationship between `Update.swift` and `FreeformPost`. Can we conform to GraphIDBridging like we do in the other graph schema classes so we get the `.graphID` property for use here?

File: Library/ViewModels/CommentsViewModel.swift [hunk 4]
```diff
@@ -216,13 +216,30 @@ public final class CommentsViewModel: CommentsViewModelType,
...
+        }
+      }
+
+    let postFailableCommentConfigData:
+      Signal<(project: Project, commentableId: String, user: User), Never> = Signal.combineLatest(
+        initialProject,
+        commentableId,
+        currentUser.skipNil()
+      ).map { configData in
```
Comment (singhhari) [inline Library/ViewModels/CommentsViewModel.swift:234]: replace with:

`.map { project, commentableId, user in
(project: project, commentableId: commentableId, user: user)
}` for readability.

File: Library/ViewModels/CommentsViewModel.swift [hunk 5]
```diff
@@ -216,13 +216,30 @@ public final class CommentsViewModel: CommentsViewModelType,
...
+        (project: configData.0, commentableId: configData.1, user: configData.2)
+      }
+
+    self.failableOrComment <~ postFailableCommentConfigData
+      .takePairWhen(commentComposerDidSubmitText)
+      .map { data in
+        (data.0.project, data.0.commentableId, data.0.user, data.1)
```
Comment (singhhari) [inline Library/ViewModels/CommentsViewModel.swift:241]: I think you can decompose this a bit for readability as well. Kinda like:

      `.map { data in
        let ((project, commentableId, user), text) = data
        return (project, commentableId, user, text)
      }`

**Review by singhhari — CHANGES_REQUESTED** [review]

File: Library/ViewModels/CommentsViewModel.swift (hunk 3 above)
Comment (moderateepheezy) [reply Library/ViewModels/CommentsViewModel.swift:225]: So we can use `GraphIDBridging` however the modelName would be `FreeformPost` , the only relationship between `Update` and `FreeformPost` is that the response to fetch comments for Update is wrapped inside `FreeformPost`.

**Review by msadoon — COMMENTED** [review]
Codewise seems fine, may want to double check failable and retry functionality carefully between project and update comments to ensure no regression.

**Review by msadoon — APPROVED** [review]
Tested basic posting between project and updates in app, may want to double-check retryable comments (the multiple failed comments case) but otherwise looks good to me.

**Review by singhhari — APPROVED** [review]
lgtm

**Pull request merged**